metricflow/model/semantics/semantic_model_lookup.py

```diff
@@ -141,6 +141,19 @@ def get_measure(self, measure_reference: MeasureReference) -> Measure:  # noqa:
             semantic_model=semantic_model, measure_reference=measure_reference
         )
 
+    def get_measure_with_agg_time_dimension(self, measure_reference: MeasureReference) -> Measure:  # noqa: D
+        measure = self.get_measure(measure_reference=measure_reference)
+        if not measure.agg_time_dimension:
+            agg_time_dimension = self.get_agg_time_dimension_for_measure(measure_reference=measure_reference)
+            if agg_time_dimension:
+                # Copy original Measure object, updating the `agg_time_dimension` field.
+                # Note: `Measure`` protocol does not include Pydantic methods.
+                # Could remove this `type: ignore` if we added the `copy` method to the protocol in DSI.
+                return measure.copy(  # type: ignore[attr-defined]
```

Comment:
We need to avoid typechecker overrides on internal operations. The contract for these methods is that the type spec is as defined, this inserts an implementation detail that will throw runtime errors for people implementing against the type spec.

If we need this method I'd rather a standard Python copy.copy() -> modify, assuming Pydantic allows it, but I think it'll be easier overall to follow the pattern we use elsewhere of having the engine return objects that are not directly sourced from here. If we do that for now we can add a copy() method to the interface later on and consolidate those secondary object types all at once.

Reply:
The reason I did it this way is that I get a type error if I try to modify that attribute of a Pydantic object:
`Property "agg_time_dimension" defined in "Measure" is read-only`
So I looked up the recommended way to do this in Pydantic.

Reply:
> I think it'll be easier overall to follow the pattern we use elsewhere of having the engine return objects that are not directly sourced from here.
The reason this doesn't seem like the right solution to me is that there isn't actually anything special or different about the `Measure` object I want to expose. To me it just seems like a bug that the `agg_time_dimension` attribute isn't populated.
The argument against that is if we want the `PydanticMeasure` object to match its YAML representation exactly - so if it's supposed to inherit the default `agg_time_dimension`, that wouldn't show up. I can implement it that way, but it seems very bug-prone to me since any developer who doesn't know that might use `PydanticMeasure` objects with the assumption that the `agg_time_dimension` field is populated with its default (as I did when I jumped into this code).

Reply:
I think this conversation is also resolved if we expose a new engine-level dataclass 👍